### PR TITLE
[sanitizer] Fix running sanitizer_bad_report_path_test on Linux as root

### PR DESCRIPTION
**Pull request by iii-i**
Running tests as root is not the greatest idea, however, there is one
    valid use case - running them in a container in order to verify LLVM on
    different distros. There is no reason to configure unprivileged users
    in a container, so one works as root.
    
    sanitizer_bad_report_path_test assumes that creating a file in a
    non-writable directory would fail, which is not the always case. For
    example, when we are on Linux and CAP_DAC_OVERRIDE, which root has, is
    in effect. Therefore, one solution is to drop it. However, that would
    be Linux-specific.
    
    Instead, use argv[0] as if it were a directory. mkdir() on top of a
    file should be prohibited by all supported Posix operating systems.
    Combine this with a partial revert of commit f4214e1469ad ("[sanitizer]
    Skip test on Android where chmod is not working"), since we shouldn't
    need to exclude Android anymore.

### CODE REVIEW

File: compiler-rt/test/sanitizer_common/TestCases/Posix/sanitizer_bad_report_path_test.cpp [hunk 1]
```diff
@@ -13,9 +13,32 @@
...
+#else
+static void try_drop_cap_dac_override(void) {}
+#endif
+
 volatile int *null = 0;
 
 int main(int argc, char **argv) {
+  try_drop_cap_dac_override();
   char buff[1000];
   sprintf(buff, "%s.report_path/report", argv[0]);
```
[the remark this answers is not in the document]
Comment (iii-i) [reply compiler-rt/test/sanitizer_common/TestCases/Posix/sanitizer_bad_report_path_test.cpp:43]: `__sanitizer_set_report_path()` creates all missing parent directories (`ReportFile::SetReportPath()` -> `RecursiveCreateParentDirs()`).

[the remark this answers is not in the document]
Comment (iii-i) [reply compiler-rt/test/sanitizer_common/TestCases/Posix/sanitizer_bad_report_path_test.cpp:43]: Come to think of it, we could try creating `/proc/self/xyz`. I would expect `/proc` to be special on BSD too. Alternatively, `argv[0]` - but there I'm not sure if we get `EBUSY` everywhere. What do you think?

[the remark this answers is not in the document]
Comment (iii-i) [reply compiler-rt/test/sanitizer_common/TestCases/Posix/sanitizer_bad_report_path_test.cpp:43]: Turns out `/proc` is deprecated in FreeBSD (https://man.freebsd.org/cgi/man.cgi?query=procfs); v14 doesn't even mount it by default.

[the remark this answers is not in the document]
Comment (iii-i) [reply compiler-rt/test/sanitizer_common/TestCases/Posix/sanitizer_bad_report_path_test.cpp]: Using argv[0] as a directory seems to work everywhere. I've updated the branch and added two small improvements related to sanitizer error messages.

[the remark this answers is not in the document]
Comment (iii-i) [reply compiler-rt/test/sanitizer_common/TestCases/Posix/sanitizer_bad_report_path_test.cpp]: Hmm, I just realized that `llvm-project` supports only squash merges. I will send the improvements as separate PRs. Sorry for the noise.

**Review by vitalybuka — APPROVED** [review]
Nice

**Comment by iii-i** [comment]
Unfortunately this is now causing a buildbot failure on Android:
```
/var/lib/buildbot/sanitizer-buildbot6/sanitizer-x86_64-linux-android/build/llvm-project/compiler-rt/test/sanitizer_common/TestCases/Posix/sanitizer_set_report_path_test.cpp:27:11: error: CHECK: expected string not found in input
// CHECK: ERROR: Can't create directory: {{.*}}Posix/Output/sanitizer_set_report_path_test.cpp.tmp
          ^
<stdin>:2:243: note: scanning from here
Path /data/local/tmp/Output/var/lib/buildbot/sanitizer-buildbot6/sanitizer-x86_64-linux-android/build/compiler_rt_build_android_arm/test/sanitizer_common/asan-arm-Android/Posix/Output/sanitizer_set_report_path_test.cpp.tmp.report_path/report.27545
                                                                                                                                                                                                                                                  ^

Input file: <stdin>
Check file: /var/lib/buildbot/sanitizer-buildbot6/sanitizer-x86_64-linux-android/build/llvm-project/compiler-rt/test/sanitizer_common/TestCases/Posix/sanitizer_set_report_path_test.cpp

-dump-input=help explains the following input dump.

Input was:
<<<<<<
          1: ERROR: Can't create directory: /data/local/tmp/Output/var/lib/buildbot/sanitizer-buildbot6/sanitizer-x86_64-linux-android/build/compiler_rt_build_android_arm/test/sanitizer_common/asan-arm-Android/Posix/Output/sanitizer_set_report_path_test.cpp.tmp
          2: Path /data/local/tmp/Output/var/lib/buildbot/sanitizer-buildbot6/sanitizer-x86_64-linux-android/build/compiler_rt_build_android_arm/test/sanitizer_common/asan-arm-Android/Posix/Output/sanitizer_set_report_path_test.cpp.tmp.report_path/report.27545
check:27                                                                                                                                                                                                                                                       X~~~~~ error: no match found
```

The problem is that `Path` and `ERROR:` messages come in the wrong order. The first is `fprintf()`ed to stdout, the second is `write()`ten to stderr. The testcase merges the two file descriptors. But the interesting part is that the test calls `fflush(stdout);` in between.

I'm not sure how to investigate that `fflush()` ends up being a no-op on that builder. I propose to work around that by doing a `write(kStderrFd)` for the `Path` message.

**Comment by iii-i** [comment]
I've posted https://github.com/llvm/llvm-project/pull/99469.

**Pull request merged**